warp10/src/main/java/io/warp10/script/functions/FETCH.java

```diff
@@ -1099,4 +984,197 @@ public Object apply(WarpScriptStack stack) throws WarpScriptException {
+      // Simple case: start is a long
+      endTs = (long) end;
+    } else if (end instanceof String) {
+      // If it's a string, it may be the string representation of a Long, a ISO8601 date or 'now'.
+      if ("now".equals(start)) {
+        endTs = now;
+      } else {
+        try {
+          endTs = Long.valueOf((String) end);
+        } catch (NumberFormatException nfe) {
+          // Not string representation of a Long, try ISO8601
+          try {
+            if (SystemUtils.isJavaVersionAtLeast(JavaVersion.JAVA_1_8)) {
+              endTs = io.warp10.script.unary.TOTIMESTAMP.parseTimestamp((String) end);
+            } else {
+              endTs = fmt.parseDateTime((String) end).getMillis() * Constants.TIME_UNITS_PER_MS;
+            }
+          } catch (WarpScriptException | IllegalArgumentException e) {
+            throw new WarpScriptException("Invalid format for parameter '" + endParamName + "'.", e);
+          }
+        }
+      }
+    } else if (null != end) {
+      // If end is not null and we cannot retrieve the timestamp, throw an error.
+      throw new WarpScriptException("Invalid format for parameter '" + endParamName + "'.");
+    }
+
+    // Check that either startTs or endTs is defined.
+    if (null == startTs && null == endTs) {
+      throw new WarpScriptException("Missing either '" + startParamName + "' or '" + endParamName + "' parameter.");
+    }
+
+    // If both are defined but swapped, swap them.
+    if (null != startTs && null != endTs && startTs > endTs) {
+      long tmp = startTs;
+      startTs = endTs;
+      endTs = tmp;
+    }
+
+    //
+    // Try to use timestamp to either determine endTs or startTs if not already defined.
+    //
+
+    Long numericTimespan = null;
+
+    if (null != timespan) {
+      // Check that endTs and startTs are not both already defined.
+      if (startTs != null && endTs != null) {
+        throw new WarpScriptException("Invalid time range specification: '" + startParamName + "', '" + endParamName + "' and '" + timespanParamName + "' cannot all be defined. Only 2 out of those 3 parameters should be defined.");
+      }
+
+      //
+      // Cast or convert timespan to a numeric one.
+      //
+      if (timespan instanceof Long) {
+        numericTimespan = (Long) timespan;
+      } else if (timespan instanceof String) {
+        // If it's a string, it may be the string representation of a Long or a ISO8601 duration.
+        // Try the former then the latter if it fails.
+        try {
+          numericTimespan = Long.parseLong((String) timespan);
+        } catch (NumberFormatException nfe) {
+          // Not string representation of a Long, try ISO8601 duration.
+          try {
+            ReadWritablePeriod period = new MutablePeriod();
+
+            ISOPeriodFormat.standard().getParser().parseInto(period, (String) timespan, 0, Locale.US);
+
+            Period p = period.toPeriod();
+
+            if (p.getMonths() != 0 || p.getYears() != 0) {
```

Comment:
Actually since we do have either a start or an end, we could technically support months and years as you are not interested in determining how long the period is but simply compute an offset to an existing timestamp.

Reply:
It should be possible but it's not trivial. Noted as TODO.